Crash when moving last card in deck to board

When I got to the end of the deck and tried to move the deck card to the board, I got this error and stack trace:

```
     [  3♠  7♥  6♣  3♦]
  K♦  K♥  *   K♣  Q♥  Q♦  4♦
  Q♣  Q♠  9♥      J♠  J♣    
  J♦  J♥  8♠     10♥ 10♦    
 10♣ 10♠          9♠  9♣    
  9♦              8♦  8♥    
  8♣              7♠  7♣    
  7♦              6♦        
  6♠              5♠        
  5♦                        
0   K♠ 1
6s
     [  3♠  7♥  6♣  4♦]
  K♦  K♥  *   K♣  Q♥  Q♦    
  Q♣  Q♠  9♥      J♠  J♣    
  J♦  J♥  8♠     10♥ 10♦    
 10♣ 10♠          9♠  9♣    
  9♦              8♦  8♥    
  8♣              7♠  7♣    
  7♦              6♦        
  6♠              5♠        
  5♦                        
0   K♠ 1
d6
1
     [  3♠  7♥  6♣  4♦]
  K♦  K♥  *   K♣  Q♥  Q♦  K♠
  Q♣  Q♠  9♥      J♠  J♣    
  J♦  J♥  8♠     10♥ 10♦    
 10♣ 10♠          9♠  9♣    
  9♦              8♦  8♥    
  8♣              7♠  7♣    
  7♦              6♦        
  6♠              5♠        
  5♦                        
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
solitaire/game.Game.getCurrentCard(...)
        /Users/nfang/workspace/go-solitaire/game/game.go:35
solitaire/game.Game.DisplayCurrentCard(...)
        /Users/nfang/workspace/go-solitaire/game/game_display.go:8
solitaire/game.Game.DisplayCards({{0xc0000cb700, 0x0, 0x24}, {0xc0000de6c0, 0x7, 0x8}, {0xc00009f140, 0x4, 0x4}, 0x0, ...})
        /Users/nfang/workspace/go-solitaire/game/game_display.go:12 +0x187
solitaire/game.Game.Display({{0xc0000cb700, 0x0, 0x24}, {0xc0000de6c0, 0x7, 0x8}, {0xc00009f140, 0x4, 0x4}, 0x0, ...})
        /Users/nfang/workspace/go-solitaire/game/game_display.go:18 +0x7a
solitaire/gamemanagerapi.HandleMoves(0xc000123ed8, {{0xc000196000, 0x50, 0x8b}})
        /Users/nfang/workspace/go-solitaire/gamemanagerapi/gamemanagerapi.go:80 +0x17c5
main.main()
        /Users/nfang/workspace/go-solitaire/cmd/main_cli.go:17 +0x2dd
exit status 2
go-solitaire/cmd - (main) > 
```

--- a/game/deck/deck.go
+++ b/game/deck/deck.go
@@ -90,6 +90,7 @@ func NewDeck() Cards {
 			card, cardErr := NewCard(value, suit, false)
 			if cardErr != nil {
 				fmt.Printf("new deck not created: %s", cardErr.Error())
+				return Cards{}
 			}
 			deck = append(
 				deck,
@@ -99,6 +100,38 @@ func NewDeck() Cards {
 	}
 
 	return deck
+}
+
+func NewTestingDeck() (Cards, error) {
+	deck := Cards{}
+	for i := 13; i >= 7; i-- {
+		for _, suit := range CardSuits {
+			card, cardErr := NewCard(i, suit, true)
+			if cardErr != nil {
+				return nil, fmt.Errorf("new deck not created: %s", cardErr.Error())
+			}
+			deck = append(deck, card)
+		}
+	}
+	for _, suit := range CardSuits {
+		for _, value := range []int{3, 2, 1, 6, 5, 4} {
+			card, cardErr := NewCard(value, suit, true)
+			if cardErr != nil {
+				return nil, fmt.Errorf("new deck not created: %s", cardErr.Error())
+			}
+			deck = append(deck, card)
+		}
+	}
+	// for _, suit := range CardSuits {
+	// 	for _, value := range []int{6, 5, 4} {
+	// 		card, cardErr := NewCard(value, suit, true)
+	// 		if cardErr != nil {
+	// 			return nil, fmt.Errorf("new deck not created: %s", cardErr.Error())
+	// 		}
+	// 		deck = append(deck, card)
+	// 	}
+	// }
+	return deck, nil
 }
 
 func (d *Cards) RemoveCard(cardIndex int) Card {

--- a/game/deck/deck_test.go
+++ b/game/deck/deck_test.go
@@ -175,6 +175,20 @@ func TestNewDeck(t *testing.T) {
 	}
 }
 
+func TestNewTestingDeck(t *testing.T) {
+	deck, err := NewTestingDeck()
+	if err != nil {
+		t.Error("Expected no error.")
+	}
+	// this order makes it easy to clear out the deck and board to test edge cases.
+	deckValues := []int{13, 13, 13, 13, 12, 12, 12, 12, 11, 11, 11, 11, 10, 10, 10, 10, 9, 9, 9, 9, 8, 8, 8, 8, 7, 7, 7, 7, 3, 2, 1, 6, 5, 4, 3, 2, 1, 6, 5, 4, 3, 2, 1, 6, 5, 4, 3, 2, 1, 6, 5, 4}
+	for i, card := range deck {
+		if card.Value != deckValues[i] {
+			t.Error("Card values are not correct.", card.Value, card)
+		}
+	}
+}
+
 func TestRemoveCard(t *testing.T) {
 	deck := NewDeck()
 	deck.RemoveCard(0)

--- a/game/game.go
+++ b/game/game.go
@@ -30,9 +30,18 @@ func checkMove(card deck.Card, toCard deck.Card) bool {
 	return false
 }
 
-func (g Game) getCurrentCard() deck.Card {
+func (g Game) getCurrentCard() (deck.Card, error) {
+	if g.Cards == nil {
+		return deck.Card{}, errors.New("deck not initialized")
+	}
+	if len(g.Cards) == 0 {
+		return deck.Card{}, errors.New("no cards in the deck")
+	}
+	if g.CurrentCardIndex >= len(g.Cards) {
+		return deck.Card{}, errors.New("current card index out of range")
+	}
 
-	return g.Cards[g.CurrentCardIndex]
+	return g.Cards[g.CurrentCardIndex], nil
 }
 
 func (g *Game) pruneColumn(column int, index int) []deck.Card {
@@ -178,7 +187,11 @@ func (g *Game) SetState(gameState Game) {
 // for the user the columns are 1 indexed instead of 0 indexed.
 
 func (g Game) GetDeckMoves() []int {
-	currentCard := g.getCurrentCard()
+	currentCard, error := g.getCurrentCard()
+	if error != nil {
+		// no deck, no moves
+		return []int{}
+	}
 	moves := []int{}
 	for index, _ := range g.Board {
 		_, lastCard := g.Board.GetLastCard(index)
@@ -226,7 +239,11 @@ func (g Game) GetDeckHints() []string {
 func (g Game) GetStackHints() []string {
 	stackHints := []string{}
 	// check deck first
-	_, validDeckMove := g.GetStackMoves(g.getCurrentCard())
+	currentCard, error := g.getCurrentCard()
+	if error != nil {
+		// no current card, nothing to add to stackHints
+	}
+	_, validDeckMove := g.GetStackMoves(currentCard)
 	if validDeckMove {
 		stackHints = append(stackHints, "ds")
 	}

--- a/game/game.go
+++ b/game/game.go
@@ -40,7 +40,9 @@ func (g Game) getCurrentCard() (deck.Card, error) {
 	if g.CurrentCardIndex >= len(g.Cards) {
 		return deck.Card{}, errors.New("current card index out of range")
 	}
-
+	if g.CurrentCardIndex == -1 {
+		return deck.Card{}, nil
+	}
 	return g.Cards[g.CurrentCardIndex], nil
 }
 
@@ -53,14 +55,14 @@ func (g *Game) pruneColumn(column int, index int) []deck.Card {
 /* Exported Functions */
 
 func NewGame() Game {
-	return Game{deck.NewDeck(), board.NewBoard(), stacks.NewStacks(), 0, false}
+	return Game{deck.NewDeck(), board.NewBoard(), stacks.NewStacks(), 2, false}
 }
 
 func (g *Game) Reset() {
 	g.Cards = deck.NewDeck()
 	g.Board = board.NewBoard()
 	g.Stacks = stacks.NewStacks()
-	g.CurrentCardIndex = 0
+	g.CurrentCardIndex = 2
 }
 
 func (g1 Game) IsEqual(g2 Game) bool {
@@ -97,7 +99,7 @@ func (g *Game) DealBoard() {
 	// the first column has 1 card, the second has 2, etc.
 	cards := g.Cards
 	board := g.Board
-	currentCardIndex := 0
+	currentCardIndex := 2
 
 	for i := 0; i < NumColumns; i++ {
 		for j := i; j < NumColumns; j++ {
@@ -118,12 +120,17 @@ func (g *Game) DealBoard() {
 }
 
 func (g *Game) NextDeckCard() error {
-	if len(g.Cards) == 0 {
+	deckLength := len(g.Cards)
+	if deckLength == 0 {
 		return errors.New("no more cards in the deck")
 	}
 	g.Cards[g.CurrentCardIndex].Shown = false
-	if g.CurrentCardIndex+3 > len(g.Cards)-1 {
-		g.CurrentCardIndex = 0
+	if g.CurrentCardIndex+3 > deckLength-1 {
+		if deckLength < 3 {
+			g.CurrentCardIndex = deckLength - 1
+		} else {
+			g.CurrentCardIndex = 2
+		}
 	} else {
 		g.CurrentCardIndex += 3
 	}

--- a/game/game.go
+++ b/game/game.go
@@ -124,12 +124,17 @@ func (g *Game) NextDeckCard() error {
 	if deckLength == 0 {
 		return errors.New("no more cards in the deck")
 	}
-	g.Cards[g.CurrentCardIndex].Shown = false
-	if g.CurrentCardIndex+3 > deckLength-1 {
-		if deckLength < 3 {
-			g.CurrentCardIndex = deckLength - 1
-		} else {
+	fmt.Println("index", g.CurrentCardIndex)
+	// if there is a current card, hide it
+	if g.CurrentCardIndex >= 0 {
+		g.Cards[g.CurrentCardIndex].Shown = false
+	}
+	// if the next card is out of bounds, set the current card back to the beginning
+	if g.CurrentCardIndex == -1 || g.CurrentCardIndex+3 > deckLength-1 {
+		if 2 < deckLength-1 {
 			g.CurrentCardIndex = 2
+		} else {
+			g.CurrentCardIndex = deckLength - 1
 		}
 	} else {
 		g.CurrentCardIndex += 3
@@ -194,12 +199,12 @@ func (g *Game) SetState(gameState Game) {
 // for the user the columns are 1 indexed instead of 0 indexed.
 
 func (g Game) GetDeckMoves() []int {
+	moves := []int{}
 	currentCard, error := g.getCurrentCard()
 	if error != nil {
 		// no deck, no moves
-		return []int{}
+		return moves
 	}
-	moves := []int{}
 	for index, _ := range g.Board {
 		_, lastCard := g.Board.GetLastCard(index)
 		if checkMove(currentCard, lastCard) {

--- a/game/game_display.go
+++ b/game/game_display.go
@@ -17,7 +17,7 @@ func (g Game) DisplayCurrentCard() string {
 }
 
 func (g Game) DisplayCards() {
-	fmt.Println(g.CurrentCardIndex, g.DisplayCurrentCard(), len(g.Cards)-g.CurrentCardIndex)
+	fmt.Println(g.CurrentCardIndex+1, g.DisplayCurrentCard(), len(g.Cards)-g.CurrentCardIndex)
 }
 
 func (g Game) Display() {

--- a/game/game_display.go
+++ b/game/game_display.go
@@ -7,6 +7,10 @@ import (
 func (g Game) DisplayCurrentCard() string {
 	card, error := g.getCurrentCard()
 	if error != nil {
+
+		if error.Error() == "no cards in the deck" {
+			return ""
+		}
 		return error.Error()
 	}
 	return card.DisplayMini

--- a/game/game_display.go
+++ b/game/game_display.go
@@ -5,7 +5,11 @@ import (
 )
 
 func (g Game) DisplayCurrentCard() string {
-	return g.getCurrentCard().DisplayMini
+	card, error := g.getCurrentCard()
+	if error != nil {
+		return error.Error()
+	}
+	return card.DisplayMini
 }
 
 func (g Game) DisplayCards() {

--- a/game/game_moves.go
+++ b/game/game_moves.go
@@ -16,13 +16,16 @@ func (g *Game) MoveFromDeckToBoard(column int) error {
 	g.Cards[g.CurrentCardIndex].Shown = true
 	g.Board[column] = append(g.Board[column], g.Cards[g.CurrentCardIndex])
 	g.Cards.RemoveCard(g.CurrentCardIndex)
-	if g.CurrentCardIndex > 0 {
+	if g.CurrentCardIndex >= 0 {
 		g.CurrentCardIndex = g.CurrentCardIndex - 1
 	}
 	return nil
 }
 
 func (g *Game) MoveFromDeckToStacks() error {
+	if g.CurrentCardIndex < 0 {
+		return errors.New("no cards in the deck")
+	}
 	suitIndex, validMove := g.GetStackMoves(g.Cards[g.CurrentCardIndex])
 	if !validMove {
 		return errors.New("invalid move")
@@ -30,7 +33,7 @@ func (g *Game) MoveFromDeckToStacks() error {
 	movedCard := g.Cards.RemoveCard(g.CurrentCardIndex)
 	movedCard.Shown = true
 	g.Stacks[suitIndex] = append(g.Stacks[suitIndex], movedCard)
-	if g.CurrentCardIndex > 0 {
+	if g.CurrentCardIndex >= 0 {
 		g.CurrentCardIndex = g.CurrentCardIndex - 1
 	}
 	return nil

--- a/game/game_test.go
+++ b/game/game_test.go
@@ -1,6 +1,7 @@
 package game
 
 import (
+	"fmt"
 	"solitaire/game/deck"
 	"solitaire/testutils"
 	"testing"
@@ -12,8 +13,17 @@ func TestGetCurrentCard(t *testing.T) {
 	if err != nil {
 		t.Error("Expected no error.")
 	}
-	if card != game.Cards[0] {
-		t.Error("Expected current card to be the first card in the deck.")
+	if card != game.Cards[2] {
+		t.Error("Expected current card to be the third card in the deck.")
+	}
+
+	game.CurrentCardIndex = -1
+	card, err = game.getCurrentCard()
+	if err != nil {
+		t.Error("Expected no error.")
+	}
+	if card.Value != 0 || card.Suit != "" {
+		t.Error("Expected current card to be empty.")
 	}
 
 	game.CurrentCardIndex = 52
@@ -33,7 +43,6 @@ func TestGetCurrentCard(t *testing.T) {
 	if err.Error() != "no cards in the deck" {
 		t.Error("Expected error message not shown.")
 	}
-
 }
 
 func TestIsEqual(t *testing.T) {
@@ -71,13 +80,46 @@ func TestNewGame(t *testing.T) {
 	if len(g.Cards) != 52 {
 		t.Error("Wrong number of cards in the deck.")
 	}
+	if g.CurrentCardIndex != 2 {
+		t.Error("Expected current card index to be 2.", g.CurrentCardIndex)
+	}
 }
 
 func TestNextDeckCard(t *testing.T) {
 	g := NewGame()
 
-	g.Cards = g.Cards[:0]
 	err := g.NextDeckCard()
+	if err != nil {
+		t.Error("Expected no error.")
+	}
+	if g.CurrentCardIndex != 5 {
+		t.Error("Expected current card index to be 5.", g.CurrentCardIndex)
+	}
+
+	g.Cards = g.Cards[:2]
+	g.CurrentCardIndex = 0
+	fmt.Println(g.Cards)
+	err = g.NextDeckCard()
+	if err != nil {
+		t.Error("Expected no error.")
+	}
+	if g.CurrentCardIndex != 1 {
+		t.Error("Expected current card index to be 1 if there are only 2 cards left.", g.CurrentCardIndex)
+	}
+
+	g.Cards = g.Cards[:1]
+	g.CurrentCardIndex = 0
+	fmt.Println(g.Cards)
+	err = g.NextDeckCard()
+	if err != nil {
+		t.Error("Expected no error.")
+	}
+	if g.CurrentCardIndex != 0 {
+		t.Error("Expected current card index to be 0 if there are only 2 cards left.", g.CurrentCardIndex)
+	}
+
+	g.Cards = g.Cards[:0]
+	err = g.NextDeckCard()
 	if err.Error() != "no more cards in the deck" {
 		t.Error("Expected error message not shown.")
 	}

--- a/game/game_test.go
+++ b/game/game_test.go
@@ -1,9 +1,40 @@
 package game
 
 import (
+	"solitaire/game/deck"
 	"solitaire/testutils"
 	"testing"
 )
+
+func TestGetCurrentCard(t *testing.T) {
+	game := NewGame()
+	card, err := game.getCurrentCard()
+	if err != nil {
+		t.Error("Expected no error.")
+	}
+	if card != game.Cards[0] {
+		t.Error("Expected current card to be the first card in the deck.")
+	}
+
+	game.CurrentCardIndex = 52
+	_, err = game.getCurrentCard()
+	if err.Error() != "current card index out of range" {
+		t.Error("Expected error message not shown.")
+	}
+
+	game.Cards = nil
+	_, err = game.getCurrentCard()
+	if err.Error() != "deck not initialized" {
+		t.Error("Expected error message not shown.")
+	}
+
+	game.Cards = deck.Cards{}
+	_, err = game.getCurrentCard()
+	if err.Error() != "no cards in the deck" {
+		t.Error("Expected error message not shown.")
+	}
+
+}
 
 func TestIsEqual(t *testing.T) {
 	game1 := NewGame()

--- a/gamemanagerapi/gamemanagerapi.go
+++ b/gamemanagerapi/gamemanagerapi.go
@@ -6,6 +6,7 @@ import (
 	"strings"
 
 	"solitaire/game"
+	"solitaire/game/deck"
 	"solitaire/game/gamestates"
 
 	"golang.org/x/exp/slices"
@@ -102,6 +103,19 @@ func HandleMoves(game *game.Game, gameStates gamestates.GameStates) {
 				toColumn, _ := strconv.ParseInt(input1, 10, 32)
 				game.MoveFromColumnToColumn(int(fromColumn), int(toColumn))
 				game.Display()
+				gameStates.SaveState(*game)
+				continue
+			}
+			if input[:2] == "rt" {
+				game.Reset()
+				deck, err := deck.NewTestingDeck()
+				if err != nil {
+					fmt.Println(err)
+				}
+				game.Cards = deck
+				game.DealBoard()
+				game.Display()
+				gameStates.Reset()
 				gameStates.SaveState(*game)
 				continue
 			}

